### PR TITLE
Refactor vis containers with nested models

### PR DESCRIPTION
**Pull request by axelboc**

### CODE REVIEW

File: src/h5web/explorer/utils.test.ts [hunk 1]
```diff
@@ -44,7 +44,6 @@ describe('Explorer utilities', () => {
         children: [],
         attributes: [],
         rawLink: rootLink,
-        rawEntity: rootGroup,
```
Comment (axelboc) [inline src/h5web/explorer/utils.test.ts:47]: I can now remove `rawEntity` from the entity model.

File: src/h5web/explorer/utils.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ import { isMyGroup, isReachable } from '../providers/utils';
 function prepareEntity(
   link: HDF5HardLink | HDF5RootLink,
   parents: MyHDF5Group[]
-): Omit<MyHDF5Entity, 'kind'> {
+): Omit<MyHDF5Group | MyHDF5Dataset | MyHDF5Datatype, 'kind'> {
```
Comment (axelboc) [inline src/h5web/explorer/utils.ts:19]: Helps type check that the `id` is provided (a link doesn't have an `id`).

File: src/h5web/visualizations/containers/HeatmapVisContainer.tsx [hunk 3]
```diff
@@ -24,7 +24,7 @@ function HeatmapVisContainer(props: VisContainerProps): ReactElement {
     return <></>;
   }
 
-  return <MappedHeatmapVis value={value} dims={dims} title={entityName} />;
+  return <MappedHeatmapVis value={value} dims={dims} title={entity.name} />;
```
Comment (axelboc) [inline src/h5web/visualizations/containers/HeatmapVisContainer.tsx:27]: yay!

File: src/h5web/visualizations/containers/NxImageContainer.tsx [hunk 4]
```diff
@@ -10,13 +9,12 @@ import { assertDefined } from '../shared/utils';
...
-  assertGroup(entity);
+  assertMyGroup(entity);
 
-  const { metadata } = useContext(ProviderContext);
-  const nxDataGroup = findNxDataGroup(entity, metadata);
+  const nxDataGroup = findMyNxDataGroup(entity);
   assertDefined(nxDataGroup, 'Expected to find NXdata group');
 
-  const nxData = useNxData(nxDataGroup, metadata);
+  const nxData = useNxData(nxDataGroup);
```
Comment (axelboc) [inline src/h5web/visualizations/containers/NxImageContainer.tsx:17]: No need to access the `metadata` object any more.

File: src/h5web/visualizations/nexus/hooks.ts [hunk 5]
```diff
@@ -19,29 +8,21 @@ import {
...
-
-  return useDatasetValues(datasetIds || []);
-}
-
-export function useNxData(group: HDF5Group, metadata: HDF5Metadata): NxData {
-  const values = useLinkedDatasetValues(group);
+export function useNxData(group: MyHDF5Group): NxData {
+  const values = useDatasetValues(
+    group.children.filter(isMyDataset).map((child) => child.id)
+  );
```
Comment (axelboc) [inline src/h5web/visualizations/nexus/hooks.ts:22]: Much simpler 😄

File: src/h5web/visualizations/nexus/hooks.ts [hunk 6]
```diff
@@ -73,15 +54,16 @@ export function useNxData(group: HDF5Group, metadata: HDF5Metadata): NxData {
...
   });
 
-  const errorsLink =
-    getLink(`${signalName}_errors`, group) || getLink('errors', group);
+  const errorsDataset =
+    getChildEntity(group, `${signalName}_errors`) ||
+    getChildEntity(group, 'errors');
   const errorsValue =
-    errorsLink && isReachable(errorsLink) && values[errorsLink.id];
+    errorsDataset && isMyDataset(errorsDataset) && values[errorsDataset.id];
```
Comment (axelboc) [inline src/h5web/visualizations/nexus/hooks.ts:61]: Pretty much equivalent for now but there's probably improvement to be made.

File: src/h5web/explorer/utils.ts (hunk 2 above)
Comment (loichuder) [reply src/h5web/explorer/utils.ts:19]: In fact, nothing prevents us from attaching an `id` to links as well, making `id` required in `MyHDF5Entity`

File: src/h5web/visualizations/nexus/hooks.ts (hunk 6 above)
Comment (loichuder) [reply src/h5web/visualizations/nexus/hooks.ts:61]: Should be fine for now. If we attach `id` to `MyHDF5Entity`, `errorsDataset && values[errorsDataset.id]` will be sufficient ?

File: src/h5web/visualizations/nexus/hooks.ts [hunk 7]
```diff
@@ -73,15 +54,16 @@ export function useNxData(group: HDF5Group, metadata: HDF5Metadata): NxData {
...
   assertOptionalArray<number>(errorsValue);
 
-  const titleLink = getLink('title', group);
+  const titleDataset = getChildEntity(group, 'title');
   const titleValue =
-    titleLink && isReachable(titleLink) && values[titleLink.id];
+    titleDataset && isMyDataset(titleDataset) && values[titleDataset.id];
```
Comment (loichuder) [inline src/h5web/visualizations/nexus/hooks.ts:66]: Same remark than for `errorsDataset`

**Review by loichuder — APPROVED** [review]
Just the `id` thing to discuss.

File: src/h5web/explorer/utils.ts (hunk 2 above)
Comment (axelboc) [reply src/h5web/explorer/utils.ts:19]: As discuss, let's think about this further. I'll merge as is for now.

**Pull request merged**